src/cashing.rs

```diff
@@ -0,0 +1,141 @@
+//! Basic casher struct which stores a closure and a hashmap.
+//! The hasmap stores key value pairs representing previous
+//! function calls.
+//!
+//! When the Casher function is run, it first does a lookup
+//! to see if the value has already been calculated. If it has,
+//! it returns that value. If it hasn't, it calculates the value,
+//! adds it to the hashmap, and returns it.
+
+use std::collections::HashMap;
+
+/// The Casher struct (Memoization) stores a function and a Hashmap.
+/// The HashMap keeps track of previous input and output for the function so
+/// that it only ever has to be called once. Use for expensive functions.
+pub struct Casher<F, U, V>
+where
+    F: Fn(U) -> V,
+    U: std::cmp::Eq + std::hash::Hash + Copy,
+    V: Copy,
+{
+    calculation: F,
+    values: HashMap<U, V>,
+}
+
+impl<F, U, V> Casher<F, U, V>
+where
+    F: Fn(U) -> V,
+    U: std::cmp::Eq + std::hash::Hash + Copy,
+    V: Copy,
+{
+    /// Constuctor for the Casher
+    /// # Examples
+    /// ```
+    /// # use contest_algorithms::cashing::Casher;
+    /// let mut squared = Casher::new(|n: u32| n*n);
+    /// ```
+    pub fn new(calculation: F) -> Casher<F, U, V> {
+        Casher {
+            calculation,
+            values: HashMap::new(),
+        }
+    }
+
+    /// Performs a lookup into the hashmap to see if the value has already
+    /// been calculated. If it has, returns the value. If it has not,
+    /// calls the function, stores the value, then returns the value
+    /// # Examples
+    /// ```
+    /// # use contest_algorithms::cashing::Casher;
+    /// let mut squared = Casher::new(|n: u32| n*n);
+    ///
+    /// // This is where we call the function
+    /// let sixteen = squared.on(4);
+    /// ```
+    pub fn on(&mut self, arg: U) -> V {
```

Comment:
If the other function is `call_and_replace()`, maybe this one should be `call`?

Or if you want to be extra-fancy, it might be possible to support function-call syntax by having a `impl Fn for Casher`, though I've never done this before...